Template Pipeline: `$any`, `$event`, host binding functions, source maps, compat mode, in-phase binding specialization, and more

dylhunn opened the pull request:
Wide-ranging PR on the template pipeline. Particular highlights include:

- We can now compile a number of host binding functions (although not all)
- For some input AST, source maps are now preserved
- Special template features `$event` and `$any` are supported
- A unified compatibility mode concept
- Major refactors to simplify binding processing across the board
- Major refactors of the compilation unit abstractions

See individual commit messages for details.

dylhunn committed:
refactor(compiler): Support `$any` in template pipeline

`$any(...)` casts should be dropped, except when they are an explicit call on `this.$any(...)`. Fix a bug in which we were transforming `ThisReceiver` into an implicit receiver.

dylhunn committed:
refactor(compiler): Introduce source maps in the template pipeline.

Begin producing source maps for the template pipeline, for a couple fundamental kinds of instructions, including elements, templates, properties, text, and interpolations.

dylhunn committed:
refactor(compiler): Fix `$event` in listeners for template pipeline

Previously, `$event` was interpreted as a lexical read on the enclosing context. Now, a new pass converts such reads into simple output AST reads of `$event`, so they are not processed by the context resolution or naming phases. Additionally, the same pass sets a field on the enclosing listener op, so that the reify phase does not have to search for reads of `$event`.

dylhunn committed:
refactor(compiler): Compatibility mode for the template pipeline.

Add a compatibility setting to the component compilation. Accordingly, remove all the custom compatibility flags passed to each phase, and use the main setting instead.

dylhunn committed:
refactor(compiler): Prepare the template pipeline to support host bindings.

Refactor `compilation.ts` by introducing two new concepts:
1. A compilation unit, which has create and update ops. Compilations of individual views are compilation units, as are individual host bindings.
2. Aa compilation job, which has several compilation units. For example, a whole component is a compilation job, because it can have many view compilation units. A host binding compilation is a job in addition to a unit, because each host binding unit is always a singleton.

Then, we begin modifying phases to accept general compilation jobs instead of component compilations specifically, which will allow us to run them on host bindings. In particular, we update the following phases: `phaseReify`, and `phaseChaining`.

dylhunn committed:
refactor(compiler): Template pipeline: accept host bindings in additional phases

Modify most of the remaining necessary phases to accept generic `CompilationJobs`. This includes `phasePureLiteralStructures`, `phaseNullishCoalescing`, `phaseExpandSafeReads`, `phaseVariableOptimization`, `phaseNaming`, and `phasePureFunctionExtraction`.

dylhunn committed:
refactor(compiler): Allow host bindings to be ingested into the template pipeline.

Alter the compiler code to ingest and process host bindings, using the newly updated compilation passes.

This is currently switched off in the outer compiler layer, but lays the foundation for actually generating the host binding functions using template pipeline.

dylhunn committed:
refactor(compiler): Refactor processing of bindings and interpolations.

Today, bindings on templates are ingested in highly distinct ways, depending on the parsed binding kind, as well as special cases for `style` and `class`. This makes it very difficult to also ingest them for host bindings without duplicating all this subtle logic.

To solve this, we introduce two major related refactors:

1. Move all processing of attributes into phases. This dramatically reduces the amount of code in `ingest.ts`, which is now only responsible for ingesting an abstract `BindingOp`. The later phases replace each `BindingOp` with more specific ops for each binding kind. For example, `binding_specialization.ts` transforms each abstract `BindingOp` into a concrete `PropertyOp`, `AttributeOp`, etc. Likewise, `style_binding_specialization.ts` performs special-case transformations for style and class bindings. This approach has the additional advantage of separating the creation of attribute and property bindings from other special cases.

2. Eliminate all interpolation ops. Instead, allow the expression inside of an op to be of a new `Interpolation` type. The reify code will then emit the appropriate instruction variant (interpolated or unary).

3. Separate some concerns that were previously mixed in, such as empty bindings and listeners on templates.

These refactors cause major downstream code changes across the system, especially to attribute extraction and variable counting.

dylhunn committed:
refactor(compiler): Allow host binding functions to specialize bindings

Interestingly, host bindings are parsed quite differently from template functions. For example, bindings such as `[style.foo]: 3px` would be parsed into a value, unit, and type when bound to a template, but will not be parsed as such when used in a host binding.

In this commit, we remedy this shortcoming by adding support for bindings in host binding functions to the template pipeline. In particular, we create a phase to process these bindings, and transform them into the correct output binding kind.

Additionally, we fix some other minor bugs and omissions.

Finally, we enable compilation of host bindings with the template pipeline, which requires us to turn off a number of failing tests.

dylhunn committed:
refactor(compiler): Enable additional tests for the template pipeline

A number of tests were previously disabled, but are now passing with the latest changes.

alxhub commented on packages/compiler/src/template/pipeline/src/phases/resolve_dollar_event.ts:
```diff
@@ -0,0 +1,37 @@
+  for (const op of ops) {
+    if (op.kind === ir.OpKind.Listener) {
+      ir.transformExpressionsInOp(op, (expr) => {
+        if (expr instanceof ir.LexicalReadExpr && expr.name === '$event') {
```
The same case occurs here with a hypothetical `<ng-template let-$event>`, but with opposite hierarchy. In this case, clearly the `$event` reference in the listener should shadow the `let` variable.

dylhunn replied:
Isn't this already what happens with the above implementation? I think this check should already pick up all lexical reads of `$event` in listeners. Or am I misunderstanding the concern?

alxhub commented on packages/compiler/src/template/pipeline/src/compilation.ts:
```diff
@@ -138,4 +142,8 @@ export class ViewCompilation {
     }
   }
+
+  get compatibility(): ir.CompatibilityMode {
```
Nit: I don't think we need this forwarding given that `.tpl.compatibility` is immediately accessible.

alxhub commented on packages/compiler/src/template/pipeline/src/compilation.ts:
```diff
@@ -10,13 +10,94 @@ import {ConstantPool} from '../../../constant_pool';
 import * as o from '../../../output/output_ast';
 import * as ir from '../ir';
 
+// A compilation unit is compiled into a template function.
```
Nit: jsdoc

alxhub commented on packages/compiler/src/template/pipeline/src/compilation.ts:
```diff
@@ -52,6 +52,12 @@ export abstract class CompilationUnit {
+   * Number of variable slots used within this view, or `null` if variables have not yet been
+   * counted.
+   */
+  vars: number|null = null;
```
This change should probably be in the commit that introduces the `CompilationJob` abstraction in the first place.

alxhub commented on packages/compiler/src/template/pipeline/src/compilation.ts:
```diff
@@ -24,6 +24,7 @@ export abstract class CompilationUnit {
    * List of update operations for this view.
    */
   readonly update = new ir.OpList<ir.UpdateOp>();
+  abstract readonly job: CompilationJob;
```
These changes could be merged into the original split into `CompilationUnit`/`CompilationJob` as well.

dylhunn replied:
This one has interference from all the call sites that were updates in subsequent commits -- let's leave it where it is, if you don't mind too much

alxhub commented on packages/compiler/src/template/pipeline/src/phases/host_style_property_parsing.ts:
```diff
@@ -0,0 +1,78 @@
+      .toLowerCase();
+}
+
+export function parseProperty(name: string): {property: string, suffix: string|null} {
```
Nit: does not need exported.

dylhunn committed:
refactor(compiler): Delete some redundant types in template pipeline.

Previously we refactored the compilation to use the concepts of "jobs" and "units." However, old type aliases were provided to avoid changing all call-sites in bulk. Here, those aliases are deleted, and call sited updated:

1. `ComponentCompilation` becomes `ComponentCompilationJob`.
2. `ViewCompilation` becomes `ViewCompilationUnit`.

dylhunn commented on packages/compiler/src/template/pipeline/src/phases/any_cast.ts:
```diff
@@ -0,0 +1,33 @@
+}
+
+function removeAnys(e: o.Expression): o.Expression {
+  if (e instanceof o.InvokeFunctionExpr && e.fn instanceof ir.LexicalReadExpr &&
```
TDB emits the following code: `i0.ɵɵtextInterpolate1(" ", ctx_r0.x, " ");`

So it's treated as a cast, and we are OK here.

alxhub commented:
This PR was merged into the repository by commit 1c553ee915b500820e11c53bffc582b65bb6b7ac.
